stdlib/public/core/StringStorageBridge.swift

```diff
@@ -335,6 +373,464 @@ extension __SharedStringStorage {
+  ) {
+    switch _contents {
+    case .native(_):
+      _getCharacters(buffer, aRange)
+    case .utf16(let arr):
+      _precondition(aRange.location >= 0 && aRange.length >= 0,
+                       "Range out of bounds")
+      _precondition(aRange.location + aRange.length <= Int(count),
+                       "Range out of bounds")
+      let range = Range(
+        uncheckedBounds: (aRange.location, aRange.location+aRange.length))
+      arr.withUnsafeBufferPointer {
+        let toCopy = UnsafeBufferPointer(rebasing: $0[range])
+        let bufPtr = UnsafeMutableBufferPointer(
+          start: buffer,
+          count: toCopy.count
+        )
+        _ = bufPtr.initialize(from: toCopy)
+      }
+    }
+  }
+
+  @objc
+  final internal var fastestEncoding: UInt {
+    @_effects(readonly) get {
+      return _getFastestEncoding()
+    }
+  }
+
+  @objc(UTF8String)
+  @_effects(readonly)
+  final internal func _utf8String() -> UnsafePointer<UInt8> {
+    // This is Cocoa's trick for returning an "autoreleased char *", but using
+    // our CoW to make it a bit faster
+    let anchor = asString._bridgeToObjectiveCImpl()
+    let unmanagedAnchor = Unmanaged.passRetained(anchor)
+    _ = unmanagedAnchor.autorelease()
+    return _cocoaStringGetUTF8Pointer(anchor)
+  }
+
+  @objc(getCString:maxLength:encoding:)
+  @_effects(releasenone)
+  final internal func getCString(
+    _ outputPtr: UnsafeMutablePointer<UInt8>, maxLength: Int, encoding: UInt
+  ) -> Int8 {
+    return _getCString(outputPtr, maxLength, encoding)
+  }
+
+  @objc(isEqualToString:)
+  @_effects(readonly)
+  final internal func isEqual(to other:AnyObject?) -> Int8 {
+    return _isEqual(other)
+  }
+
+  @objc(copyWithZone:)
+  final internal func copy(with zone: _SwiftNSZone?) -> AnyObject {
+    return _contents.copy()
+  }
+  
+  @objc(copy)
+  final internal func copy() -> AnyObject {
+    return _contents.copy()
+  }
+  
+  @objc(mutableCopyWithZone:)
+  final internal func mutableCopy(with zone: _SwiftNSZone?) -> AnyObject {
+    return _contents.mutableCopy()
+  }
+  
+  @objc(mutableCopy)
+  final internal func mutableCopy() -> AnyObject {
+    return _contents.mutableCopy()
+  }
+  
+  /*
+   NSString, unlike Swift String, does not require its contents to be valid
+   Unicode bytes. For example, you can simply replace the first character in a
+   surrogate pair with something incorrect.
+   
+   This presents awkwardness for us when people try to build up a mutable string
+   character by character, leaving it briefly invalid.
+   
+   To handle that, we have two "modes" for _SwiftNSMutableString:
+    • String-backed, where we know we have valid contents
+    • Byte-Array-backed, where we're just a buffer
+   
+   We switch from String to Byte-Array backing when we encounter either of the
+   following situations:
+    • The range of a mutation splits a unicode scalar
+    • The replacement bytes of a mutation are not themselves valid UTF-16
+   */
```

Comment:
Should this comment go higher up, on the enum?

Reply:
I'm torn. I sorta like having it near where the choices described are being made, but you're right, it does describe the type.